### PR TITLE
Restructure the image modules to dataclients

### PR DESCRIPTION
**Pull request by addityasingh**
- [x] Restructure the image modules to dataclients
- [x] Add `cargo fmt` on githook
- [x] Fix breaking type issue

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,35 +1,22 @@
-extern crate image;
-extern crate dssim;
-extern crate imgref;
-extern crate lodepng;
-use std::env;
-use std::path::{Path};
-use image::{open, ImageResult, DynamicImage};
-use imgref::{ImgVec};
-use dssim::*;
+extern crate optimizer;
```
Comment (axyz) [inline src/main.rs]: I would move main inside a bin folder and have lib directly in the source folder, so we can do
```rs
extern crate optimal_image
```
and keep the lib name consistent with the crate

what you think?

File: src/optimizer/dataclients/mod.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+use imgref::{ImgVec};
+use dssim::*;
+mod png;
+
+trait ImageSpec {
+    fn new(&self, name: String) -> Self where Self: Sized;
+    fn load<T: AsRef<Path>>(&self, path: T) -> Result<ImgVec<dssim::RGBAPLU>, io::Error>;
+}
+
+pub use self::png::Png;
```
Comment (axyz) [inline src/optimizer/dataclients/mod.rs]: is there a rustfmt config to include an empty line at the end?

File: src/optimizer/dataclients/png.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+use dataclients::ImageSpec;
+
+#[derive(Debug)]
+pub struct Png {
+    name: String
+}
+
+impl ImageSpec for Png {
+    fn new(&self, name: String) -> Png {
+        Png{ name }
```
Comment (axyz) [inline src/optimizer/dataclients/png.rs]: what is the usage of name here?

Comment (addityasingh) [reply src/optimizer/dataclients/png.rs]: Name could be the unique identifier for the specific trait implementation which can be used for logging and debugging purposes later

File: src/optimizer/dataclients/mod.rs (hunk 2 above)
Comment (addityasingh) [reply src/optimizer/dataclients/mod.rs]: This is already included by default

**Pull request merged**